src/games/preference.c

```diff
@@ -263,6 +264,10 @@ igraph_error_t igraph_preference_game(igraph_t *graph, igraph_integer_t nodes,
                 }
             }
 
+            if (maxedges > IGRAPH_MAX_EXACT_REAL) {
+                IGRAPH_ERROR("Overflow in number of edges.", IGRAPH_EOVERFLOW);
```

Comment:
This is technically not an overflow in the number of edges, but in the maximum number of edges. What would be a better error message here that also communicates to the user what they need to do (i.e. reduce the vertex count)?

Maybe just be as explicit as possible and write

```
Too many vertices, overflow in maximum number of edges.
```

What do you think @GroteGnoom ?

Reply:
I completely agree 🙂

Reply:
There were some other instances of this, which can be found by searching for `IGRAPH_MAX_EXACT_REAL`. Can you fix those too?

Also, the same type of fix will be necessary for all functions that are "G(n,p)-like". Off the top of my head, `erdos_renyi_gnp`, `sbm`, `hsbm`, `islands`. Grepping for `maxedges` will be helpful in finding these. Can you please fix these too?

Reply:
sbm seems to be a different situation, but I still think the check is correct.

Don't we have 2 problems? One for maxedges > MAX_EXACT_REAL, but also one where igraph_integer_t is 32 bits and maxedges overflows it when calling igraph_vector_reserve(.., maxedges) ?

Reply:
In G(n,p)-style generators, we never reserve a vector of size maxedges. This value is easily much larger than available memory. Where exactly do you see `maxedges` storage being reserved? Can you please post a direct link to the line?

Reply:
All instances related to `IGRAPH_MAX_EXACT_REAL` seem to be fixed in the PR.

I have searched the codebase for `igraph_.*_reserve.*maxedges` (regex) and there are four instances where we try to reserve space for a vector with `maxedges * p * 1.1` elements. On 32-bit machines this can potentially overflow if `maxedges * p * 1.1 > INT32_MAX`, but it's even worse because converting a double to an integer type when the double is out of bounds for the integer type is undefined behaviour. So I'm adding an `igraph_i_safe_ceil()` and `igraph_i_safe_floor()` to truncate the double first using `ceil()` or `floor()` (to make the intent explicit), and then do a range check before storing the result in an integer.

Reply:
Resolved in [e844432](https://github.com/igraph/igraph/pull/2097/commits/e844432ca048e47d0833fc47f51637bb94445349)